crates/pallets/starknet/src/simulations.rs

```diff
@@ -0,0 +1,134 @@
+                            "Transaction execution reverted during fee estimation: {}",
+                            // Safe due to the `match` branch order
+                            execution_info.revert_error.unwrap()
+                        );
+                        Err(Error::<T>::TransactionExecutionFailed)
+                    }
+                }
+            })
+            .map(|exec_info_res| {
+                exec_info_res.map(|exec_info| {
+                    exec_info
+                        .actual_resources
+                        .0
+                        .get("l1_gas_usage")
+                        .ok_or_else(|| DispatchError::from(Error::<T>::MissingL1GasUsage))
+                        .map(|l1_gas_usage| (exec_info.actual_fee.0 as u64, *l1_gas_usage))
+                })
+            });
+
+        let mut fees = Vec::with_capacity(transactions_len);
+        for fee_res in fee_res_iterator {
+            fees.push(fee_res??);
+        }
+
+        Ok(fees)
+    }
+
+    pub fn simulate_transactions(
+        transactions: Vec<UserTransaction>,
+        simulation_flags: &SimulationFlags,
+    ) -> Result<Vec<Result<TransactionExecutionInfo, PlaceHolderErrorTypeForFailedStarknetExecution>>, DispatchError>
+    {
+        let chain_id = Self::chain_id();
+        let block_context = Self::get_block_context();
+        let mut execution_config =
+            RuntimeExecutionConfigBuilder::new::<T>().with_simulation_mode(simulation_flags).build();
+
+        let tx_execution_results = transactions
+            .into_iter()
+            .map(|tx| {
+                execution_config.set_offset_version(tx.offset_version());
+                match tx {
+                    UserTransaction::Declare(tx, contract_class) => tx
+                        .try_into_executable::<T::SystemHash>(chain_id, contract_class.clone(), tx.offset_version())
+                        .and_then(|exec| {
+                            exec.execute(&mut BlockifierStateAdapter::<T>::default(), &block_context, &execution_config)
+                        }),
+                    UserTransaction::DeployAccount(tx) => {
+                        let executable = tx.into_executable::<T::SystemHash>(chain_id, tx.offset_version());
+                        executable.execute(
+                            &mut BlockifierStateAdapter::<T>::default(),
+                            &block_context,
+                            &execution_config,
+                        )
+                    }
+                    UserTransaction::Invoke(tx) => {
+                        let executable = tx.into_executable::<T::SystemHash>(chain_id, tx.offset_version());
+                        executable.execute(
+                            &mut BlockifierStateAdapter::<T>::default(),
+                            &block_context,
+                            &execution_config,
+                        )
+                    }
+                }
```

Comment:
let's move the common code to get exeuction_info to a function? also, i didn't understand, in this case how are the changes reverted?

Reply:
They are not. Except if you call from the RuntimeAPI, then everything you do is reverted anyway.
That's why I added those big warning comments. And we have to be extra careful not to let anyone write code that would use those from inside an extrinsic

Reply:
> let's move the common code to get exeuction_info to a function?

Done